geopandas/io/geoarrow.py

```diff
@@ -1,9 +1,334 @@
+                pa.field("z", pa.float64(), nullable=False),
+            ]
+            parr = pa.StructArray.from_arrays(
+                [coords[:, 0].copy(), coords[:, 1].copy(), coords[:, 2].copy()],
+                fields=fields,
+                mask=mask,
+            )
+    return parr
+
+
+def _linestring_type(point_type):
+    return pa.list_(pa.field("vertices", point_type, nullable=False))
+
+
+def _polygon_type(point_type):
+    return pa.list_(
+        pa.field(
+            "rings",
+            pa.list_(pa.field("vertices", point_type, nullable=False)),
+            nullable=False,
+        )
+    )
+
+
+def _multipoint_type(point_type):
+    return pa.list_(pa.field("points", point_type, nullable=False))
+
+
+def _multilinestring_type(point_type):
+    return pa.list_(
+        pa.field("linestrings", _linestring_type(point_type), nullable=False)
+    )
+
+
+def _multipolygon_type(point_type):
+    return pa.list_(pa.field("polygons", _polygon_type(point_type), nullable=False))
+
+
+def construct_geometry_array(
+    shapely_arr: NDArray[np.object_],
+    include_z: Optional[bool] = None,
+    *,
+    field_name: str = "geometry",
+    crs: Optional[str] = None,
+    interleaved: bool = True,
+) -> Tuple[pa.Field, pa.Array]:
+    # NOTE: this implementation returns a (field, array) pair so that it can set the
+    # extension metadata on the field without instantiating extension types into the
+    # global pyarrow registry
+    geom_type, coords, offsets = shapely.to_ragged_array(
+        shapely_arr, include_z=include_z
+    )
+
+    mask = shapely.is_missing(shapely_arr)
+    if mask.any():
+        if (
+            geom_type == GeometryType.POINT
+            and interleaved
+            and Version(pa.__version__) < Version("15.0.0")
+        ):
+            raise ValueError(
+                "Converting point geometries with missing values is not supported "
+                "for interleaved coordinates with pyarrow < 15.0.0. Please "
+                "upgrade to a newer version of pyarrow."
+            )
+        mask = pa.array(mask, type=pa.bool_())
+
+        if geom_type == GeometryType.POINT and not SHAPELY_GE_204:
+            # bug in shapely < 2.0.4, see https://github.com/shapely/shapely/pull/2034
+            # this workaround only works if there are no empty points
+            indices = np.nonzero(mask)[0]
+            indices = indices - np.arange(len(indices))
+            coords = np.insert(coords, indices, np.nan, axis=0)
+
+    else:
+        mask = None
+
+    if coords.shape[-1] == 2:
+        dims = "xy"
+    elif coords.shape[-1] == 3:
+        dims = "xyz"
+    else:
+        raise ValueError(f"Unexpected coords dimensions: {coords.shape}")
+
+    extension_metadata: Dict[str, str] = {}
+    if crs is not None:
+        extension_metadata["ARROW:extension:metadata"] = json.dumps({"crs": crs})
+
+    if geom_type == GeometryType.POINT:
+        # TODO support shapely < 2.0.4 for missing values
```

Comment:
Is this a stale comment, handled by [lines 215-220](https://github.com/geopandas/geopandas/pull/3219/files#diff-641f9c81b32c5622e37551d50ad3e74e58ae15c8378b610b55d9cc6160b23b60R215-R220)?